src/metabase/query_processor/middleware/fetch_source_query.clj

```diff
@@ -44,262 +20,132 @@
+                      {:type qp.error-type/disabled-feature, :card-id (:source-card first-stage)})))
+    ;; If the first stage came from a different source card (i.e., we are doing recursive resolution) record the
+    ;; dependency of the previously-resolved source card on the one we're about to resolve. We can check for circular
+    ;; dependencies this way.
+    (when (:qp/stage-is-from-source-card first-stage)
+      (u/prog1 (swap! dep-graph
+                      dep/depend
+                      (tru "Card {0}" (:qp/stage-is-from-source-card first-stage))
+                      (tru "Card {0}" (:source-card first-stage)))
+        ;; This will throw if there's a cycle
+        (dep/topo-sort <>)))
+    (let [card         (card query (:source-card first-stage))
+          card-stages  (get-in card [:dataset-query :stages])
+          ;; this information is used by [[metabase.query-processor.middleware.annotate/col-info-for-field-clause*]]
+          first-stage' (-> first-stage
+                           (assoc :qp/stage-had-source-card (:id card))
+                           (dissoc :source-card))
+          stages'      (into (vec card-stages)
+                             cat
+                             [[first-stage']
+                              (rest stages)])]
+      {:stages stages', :card-id (:id card)})))
+
+(defn- resolve-source-cards-in-joins [query dep-graph]
+  (lib.walk/walk
+   query
+   (fn [query path-type path join]
+     (when (= path-type :lib.walk/join)
+       (when-let [{stages' :stages} (resolve-source-cards-in-stages
+                                     query
+                                     (get-in query (conj (vec path) :stages))
+                                     dep-graph)]
+         (assoc join :stages stages'))))))
+
+(def ^:private max-recursion-depth 50)
+
+(defn- resolve-source-cards* [original-query recursion-depth dep-graph]
+  ;; this is mostly to catch programmer bugs and avoid infinite loops, thus not i18n'ed. Dep graph should circular
+  ;; dependencies if they occur
+  (assert (<= recursion-depth max-recursion-depth)
+          (format "Source Cards not fully resolved after %d iterations." max-recursion-depth))
+  (let [updated-query                              (resolve-source-cards-in-joins original-query dep-graph)
+        {updated-stages :stages, card-id :card-id} (resolve-source-cards-in-stages updated-query (:stages updated-query) dep-graph)
+        ;; `:qp/source-card-id` is used by [[metabase.query-processor.middleware.results-metadata/record-metadata!]] to
+        ;; decide whether to record metadata as well as by the [[add-dataset-info]] post-processing middleware, and
+        ;; by [[metabase.query-processor.middleware.permissions/check-query-permissions*]]
+        updated-query                              (cond-> updated-query
+                                                     updated-stages (assoc :stages updated-stages)
```

Comment:
It would be nice if the new `walk` util had an ability to replace a single stage with multiple, like macro expansion `~@body`

Reply:
That's a good point, that would be useful. I'll see if I can add that